src/main/solution_space/path_finder/path_finder_v_3.py

```diff
@@ -23,99 +25,110 @@ def find_next_vertex(self, user_vertex: Vertex) -> Vertex:
-            log.info('Choosing best vertex from found vertex children')
-            return self.__choose_best_vertex(user_vertex, vertex_in_graph.children)
-
-        # We want to have TOP_N_CANON trees with nodes number, that is close to user_nodes_number.
-        # So we consequently add vertices with node numbers equal:
-        # 1. user_nodes_number
-        # 2. user_nodes_number - 1, user_nodes_number + 1
-        # 3. user_nodes_number - 2, user_nodes_number + 2
-        # 4. ....
-        # until we reach TOP_N_CANON or have no more node_numbers to add
-
-        user_canon_nodes_number = get_nodes_number_in_ast(user_vertex.canon_tree)
-        candidates = []
-        nodes_numbers = collections.deque([user_canon_nodes_number])
-        indent = 0
-        while len(candidates) < TOP_N_CANON and nodes_numbers:
-            while nodes_numbers:
-                nodes_number = nodes_numbers.pop()
-                candidates += [Vertex.get_item_by_id(id) for id in self._graph.canon_trees_nodes_number[nodes_number]]
-            indent += 1
-            lower_bound = user_canon_nodes_number - indent
-            upper_bound = user_canon_nodes_number + indent
-            if lower_bound >= 0:
-                nodes_numbers.append(lower_bound)
-            if upper_bound <= max(self._graph.canon_trees_nodes_number.keys()):
-                nodes_numbers.append(upper_bound)
-
-        # Todo: also choose anon_trees?
-
-        return self.__choose_best_vertex(user_vertex, candidates)
+        if graph_vertex:
+            graph_anon_tree = graph_vertex.serialized_code.find_anon_tree(user_anon_tree.tree)
+            if graph_anon_tree:
+                log.info('Found the same anon tree')
+                # Todo: what to do if there is the exact vertex? Find a path to the goal? Now just returning it
+                return graph_anon_tree
+
+        canon_nodes_number = get_nodes_number_in_ast(user_canon_tree)
+        anon_nodes_number = get_nodes_number_in_ast(user_anon_tree.tree)
+
+        # Get vertices ids with canon trees, which have nodes number similar to user canon_nodes_number
+        vertices_ids = self.__get_top_n_candidates(CANON_TOP_N, canon_nodes_number, self._graph.canon_trees_nodes_number)
+        vertices: List[Vertex] = [Vertex.get_item_by_id(id) for id in vertices_ids]
+
+        anon_trees = sum([v.serialized_code.anon_trees for v in vertices], [])
+        anon_nodes_numbers_dict = {k: list(v) for k, v in groupby(anon_trees, lambda a_t: a_t.nodes_number)}
+        anon_candidates = self.__get_top_n_candidates(ANON_TOP_N, anon_nodes_number, anon_nodes_numbers_dict)
+
+        return self.__choose_best_anon_tree(user_anon_tree, anon_candidates)
 
     @staticmethod
-    def __get_rollback_probability(user_canon_tree: ast.AST, vertex_canon_tree: ast.AST) -> float:
+    def __get_top_n_candidates(top_n: int, nodes_number: int, nodes_numbers_dict: Dict[int, List[Any]]) -> List[Any]:
```

Comment:
Please, add **todo** for the method, because if we have sparse data it will work very long. We can use a modified binary search for example